// @concordium/web-sdk/lib/esm/grpc-api/v2/concordium/protocol-level-tokens.js
// @generated by protobuf-ts 2.9.1 with parameter optimize_code_size
// @generated from protobuf file "v2/concordium/protocol-level-tokens.proto" (package "concordium.v2.plt", syntax proto3)
// tslint:disable
import { MessageType } from "@protobuf-ts/runtime";
import { Memo } from "./kernel.js";
import { AccountAddress } from "./kernel.js";
// @generated message type with reflection information, may provide speed optimized methods
class CBor$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.CBor", [
            { no: 1, name: "value", kind: "scalar", T: 12 /*ScalarType.BYTES*/ }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.CBor
 */
export const CBor = new CBor$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenId$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenId", [
            { no: 1, name: "value", kind: "scalar", T: 9 /*ScalarType.STRING*/ }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenId
 */
export const TokenId = new TokenId$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenModuleRef$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenModuleRef", [
            { no: 1, name: "value", kind: "scalar", T: 12 /*ScalarType.BYTES*/ }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenModuleRef
 */
export const TokenModuleRef = new TokenModuleRef$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenAmount$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenAmount", [
            { no: 1, name: "value", kind: "scalar", T: 4 /*ScalarType.UINT64*/, L: 0 /*LongType.BIGINT*/ },
            { no: 2, name: "decimals", kind: "scalar", T: 13 /*ScalarType.UINT32*/ }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenAmount
 */
export const TokenAmount = new TokenAmount$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenState$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenState", [
            { no: 1, name: "token_module_ref", kind: "message", T: () => TokenModuleRef },
            { no: 3, name: "decimals", kind: "scalar", T: 13 /*ScalarType.UINT32*/ },
            { no: 4, name: "total_supply", kind: "message", T: () => TokenAmount },
            { no: 5, name: "module_state", kind: "message", T: () => CBor }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenState
 */
export const TokenState = new TokenState$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenAccountState$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenAccountState", [
            { no: 1, name: "balance", kind: "message", T: () => TokenAmount },
            { no: 4, name: "module_state", kind: "message", T: () => CBor }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenAccountState
 */
export const TokenAccountState = new TokenAccountState$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenModuleEvent$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenModuleEvent", [
            { no: 1, name: "type", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: "details", kind: "message", T: () => CBor }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenModuleEvent
 */
export const TokenModuleEvent = new TokenModuleEvent$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenHolder$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenHolder", [
            { no: 1, name: "account", kind: "message", oneof: "address", T: () => AccountAddress }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenHolder
 */
export const TokenHolder = new TokenHolder$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenTransferEvent$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenTransferEvent", [
            { no: 1, name: "from", kind: "message", T: () => TokenHolder },
            { no: 2, name: "to", kind: "message", T: () => TokenHolder },
            { no: 3, name: "amount", kind: "message", T: () => TokenAmount },
            { no: 4, name: "memo", kind: "message", T: () => Memo }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenTransferEvent
 */
export const TokenTransferEvent = new TokenTransferEvent$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenSupplyUpdateEvent$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenSupplyUpdateEvent", [
            { no: 1, name: "target", kind: "message", T: () => TokenHolder },
            { no: 2, name: "amount", kind: "message", T: () => TokenAmount }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenSupplyUpdateEvent
 */
export const TokenSupplyUpdateEvent = new TokenSupplyUpdateEvent$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenEvent$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenEvent", [
            { no: 1, name: "token_id", kind: "message", T: () => TokenId },
            { no: 2, name: "module_event", kind: "message", oneof: "event", T: () => TokenModuleEvent },
            { no: 3, name: "transfer_event", kind: "message", oneof: "event", T: () => TokenTransferEvent },
            { no: 4, name: "mint_event", kind: "message", oneof: "event", T: () => TokenSupplyUpdateEvent },
            { no: 5, name: "burn_event", kind: "message", oneof: "event", T: () => TokenSupplyUpdateEvent }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenEvent
 */
export const TokenEvent = new TokenEvent$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenEffect$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenEffect", [
            { no: 1, name: "events", kind: "message", repeat: 1 /*RepeatType.PACKED*/, T: () => TokenEvent }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenEffect
 */
export const TokenEffect = new TokenEffect$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenModuleRejectReason$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenModuleRejectReason", [
            { no: 1, name: "token_id", kind: "message", T: () => TokenId },
            { no: 2, name: "type", kind: "scalar", T: 9 /*ScalarType.STRING*/ },
            { no: 3, name: "details", kind: "message", T: () => CBor }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenModuleRejectReason
 */
export const TokenModuleRejectReason = new TokenModuleRejectReason$Type();
// @generated message type with reflection information, may provide speed optimized methods
class CreatePLT$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.CreatePLT", [
            { no: 1, name: "token_id", kind: "message", T: () => TokenId },
            { no: 2, name: "token_module", kind: "message", T: () => TokenModuleRef },
            { no: 4, name: "decimals", kind: "scalar", T: 13 /*ScalarType.UINT32*/ },
            { no: 5, name: "initialization_parameters", kind: "message", T: () => CBor }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.CreatePLT
 */
export const CreatePLT = new CreatePLT$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TokenCreationDetails$Type extends MessageType {
    constructor() {
        super("concordium.v2.plt.TokenCreationDetails", [
            { no: 1, name: "create_plt", kind: "message", T: () => CreatePLT },
            { no: 2, name: "events", kind: "message", repeat: 1 /*RepeatType.PACKED*/, T: () => TokenEvent }
        ]);
    }
}
/**
 * @generated MessageType for protobuf message concordium.v2.plt.TokenCreationDetails
 */
export const TokenCreationDetails = new TokenCreationDetails$Type();
